docs/source/notes/cuda.rst

```diff
@@ -635,6 +635,201 @@ of the alloc/free functions that match the signatures specified above.
 
 .. cublas-workspaces:
 
+Mixing different CUDA system allocators in the same program
+-----------------------------------------------------------
+Depending on your use case, :func:`torch.cuda.memory.change_current_allocator` may not be what you
+want to use, since it swaps the CUDA allocator for the entire program (similar to
+`PYTORCH_CUDA_ALLOC_CONF=backend:cudaMallocAsync`). For instance, if the swapped allocator doesn't
+have caching mechanism, you will lose all the benefits of PyTorch's CUDACachingAllocator. Instead,
+you can selectively mark a region of PyTorch code to use a custom allocator using
+:class:`torch.cuda.MemPool`. This will let you use multiple CUDA system allocators in the same
+PyTorch program, along with most of the benefits of the CUDACachingAllocator (e.g. caching).
+Using :class:`torch.cuda.MemPool`, you can enable several features, such as:
+
+* Allocating output buffers for an all-reduce using ``ncclMemAlloc`` allocator can enable NVLink
+  Switch Reductions (NVLS). This can reduce contention between overlapping compute and communication
+  kernels on GPU resources (SMs, and Copy Engines), especially on tensor-parallel workloads.
+* Allocating host outputs buffers for an all-gather using ``cuMemCreate`` and specifying
+  ``CU_MEM_LOCATION_TYPE_HOST_NUMA`` can enable Extended GPU Memory (EGM) based memory transfers
+  from source GPUs to the destination CPU. This accelerates the all-gather since the transfer
+  happens over NVLinks, which otherwise would have happened over bandwidth-limited, Network Interface
+  Card (NIC) links. Such an accelerated all-gather can in turn speed up model checkpointing.
+* If you are crafting a model and don't want to think about the optimal memory placements of a memory
+  intensive module at first (e.g. an embedding table), you could allocate that module with
```

Comment:
maybe add the caveat that this isn't for cases that are performance sensitive, perhaps where you just want to avoid OOM so it better agrees with the next point